File: src/PoEformance.Game/Entities/MonsterVarieties.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PoEformance.Game.Entities;

/// <summary>
/// What the game's own table says about one kind of monster.
/// </summary>
/// <remarks>
/// UNITS ARE STATED ONLY WHERE THE NUMBERS SETTLE THEM. Across 2733 rows
/// <see cref="Life"/>, <see cref="Damage"/>, <see cref="Xp"/> and <see cref="ModelSize"/> sit at
/// a median of 100-115 and run to a few hundred, which is a PERCENTAGE with 100 as the baseline.
/// <see cref="AttackSpeed"/> (0..7170, median 1500), <see cref="Speed"/> (0..97, median 24) and
/// the aggro ranges (0..500) are left unlabelled on purpose: they are plainly not percentages,
/// and naming a unit this table cannot prove is how a display ends up confidently wrong.
///
/// <see cref="Crit"/> is the trap in that set. It is called AttackCrit and holds 0, 1 or 2 - it
/// is a kind, not a chance - so anything drawing it as "0%" would be inventing a statistic.
///
/// EVERY FIELD HERE THAT IS A ROW NUMBER STAYS ONE, even the resolved ones. <see cref="Type"/>,
/// <see cref="Tags"/>, <see cref="Effects"/>, <see cref="Mods"/>, <see cref="Mods2"/> and
/// <see cref="SpecialMods"/> now have tables to resolve against, and the names live beside them
/// on <see cref="MonsterVarieties"/> rather than replacing them - the number is the game's own
/// identity, and the next table keyed on it joins without another export.
/// <see cref="Quest"/> is the last one with nowhere to point, and always will be: QuestFlags is
/// read from the game's own memory rather than shipped, so it resolves at runtime or not at all.
/// </remarks>
/// <param name="Name">
/// What the game calls it. Absent on the 24 rows whose name is a bracketed placeholder.
/// </param>
/// <param name="Quest">
/// The QuestFlags row this monster's death sets, on the 68 rows that have one - every named
/// campaign boss and nothing else. See <see cref="MonsterVarieties"/> for what that is worth.
/// </param>
/// <param name="Boss">Whether the game gives it the boss health bar. True on 363 rows.</param>
/// <param name="Inherits">
/// Object templates this monster is built on. NOT other rows of this table: of 810 references
/// exactly one names another monster variety, and the rest are .ot files - a different system.
/// The most-referenced ones are league bases (AbyssMonsterBase, SanctumMonsterBase,
/// UltimatumMonsterBase), which is what makes this the interesting column despite that.
/// </param>
public sealed record MonsterVariety(
    [property: JsonPropertyName("name")] string? Name = null,
    [property: JsonPropertyName("type")] int Type = 0,
    [property: JsonPropertyName("speed")] int Speed = 0,
    [property: JsonPropertyName("size")] int Size = 0,
    [property: JsonPropertyName("modelSize")] int ModelSize = 0,
    [property: JsonPropertyName("minAttack")] int MinAttack = 0,
    [property: JsonPropertyName("maxAttack")] int MaxAttack = 0,
    [property: JsonPropertyName("minAggro")] int MinAggro = 0,
    [property: JsonPropertyName("maxAggro")] int MaxAggro = 0,
    [property: JsonPropertyName("xp")] int Xp = 0,
    [property: JsonPropertyName("damage")] int Damage = 0,
    [property: JsonPropertyName("life")] int Life = 0,
    [property: JsonPropertyName("attackSpeed")] int AttackSpeed = 0,
    [property: JsonPropertyName("crit")] int Crit = 0,
    [property: JsonPropertyName("blood")] int Blood = 0,
    [property: JsonPropertyName("quest")] int Quest = 0,
    [property: JsonPropertyName("poise")] double Poise = 0d,
    [property: JsonPropertyName("stance")] string? Stance = null,
    [property: JsonPropertyName("boss")] bool Boss = false,
    [property: JsonPropertyName("base")] string? Base = null,
    [property: JsonPropertyName("tags")] IReadOnlyList<int>? Tags = null,
    [property: JsonPropertyName("effects")] IReadOnlyList<int>? Effects = null,
    [property: JsonPropertyName("mods")] IReadOnlyList<int>? Mods = null,
    [property: JsonPropertyName("mods2")] IReadOnlyList<int>? Mods2 = null,
    [property: JsonPropertyName("specialMods")] IReadOnlyList<int>? SpecialMods = null,
    [property: JsonPropertyName("inherits")] IReadOnlyList<string>? Inherits = null)
{
    /// <summary>How many skills the game grants this monster.</summary>
    /// <remarks>
    /// The count is worth showing even while the names are not available: it separates a
    /// monster with one attack from a boss with sixty-seven, which is the difference between
    /// something to walk past and something to look at.
    /// </remarks>
    public int SkillCount => Effects?.Count ?? 0;

    /// <summary>Whether the table says anything about this monster beyond its existence.</summary>
    public bool SaysSomething => Name is { Length: > 0 } || SkillCount > 0;
}

/// <summary>One stat a modifier sets, and the range it sets it to.</summary>
/// <remarks>
/// The export writes every value as a [min, max] pair and almost all of them are the same number
/// twice. Both are kept anyway: a modifier that really does roll is not distinguishable from a
/// fixed one once the second half is thrown away, and there is no second export to go back to.
/// </remarks>
public sealed record ModifierStat(
    [property: JsonPropertyName("stat")] string Stat = "",
    [property: JsonPropertyName("min")] int Min = 0,
    [property: JsonPropertyName("max")] int Max = 0)
{
    /// <summary>The range as one readable value.</summary>
    public string Range => Min == Max
        ? Min.ToString(System.Globalization.CultureInfo.InvariantCulture)
        : $"{Min.ToString(System.Globalization.CultureInfo.InvariantCulture)}"
          + $"-{Max.ToString(System.Globalization.CultureInfo.InvariantCulture)}";
}

/// <summary>What one of a monster's modifier rows is called, and what it does.</summary>
public sealed record ModifierMeaning(
    [property: JsonPropertyName("id")] string Id = "",
    [property: JsonPropertyName("stats")] IReadOnlyList<ModifierStat>? Stats = null);

/// <summary>
/// The defensive block the game hangs on a monster's TYPE rather than on the monster.
/// </summary>
/// <remarks>
/// THIS IS WHERE ARMOUR ACTUALLY LIVES. MonsterVarieties has a MonsterArmour column of its own
/// and it is filled on 16 rows of 2734 - reading it and concluding the game does not store
/// monster armour would be wrong twice over. The real values sit one join away, on the type:
/// Armour on 44% of the types in use, Evasion on 30%, EnergyShieldFromLife on 23%.
///
/// <see cref="Resistances"/> stays a row number, resolved through
/// <see cref="MonsterVarieties.ResistancesOf"/> - and it is a profile NAME rather than a
/// percentage, because the numbers behind it are 32 columns of tiers this export cannot explain.
/// </remarks>
/// <param name="Id">
/// What the type is called - LumberingDead, YamaBoss, IgnagdukBogWitch. It names the monster,
/// which is what makes the alignment of this table checkable at all.
/// </param>
public sealed record MonsterKind(
    [property: JsonPropertyName("id")] string Id = "",
    [property: JsonPropertyName("armour")] int Armour = 0,
    [property: JsonPropertyName("evasion")] int Evasion = 0,
    [property: JsonPropertyName("energyShield")] int EnergyShield = 0,
    [property: JsonPropertyName("spread")] int Spread = 0,
    [property: JsonPropertyName("summoned")] bool Summoned = false,
    [property: JsonPropertyName("resistances")] IReadOnlyList<int>? Resistances = null);

/// <summary>
/// The game's monster table, keyed by the path an entity carries.
/// </summary>
/// <remarks>
/// THE ENTITY'S PATH IS THE TABLE'S KEY, which is what makes this worth loading at all: no
/// fuzzy matching, no name cleanup, the Id column IS <c>Entity.Path</c>. Measured against 21
/// captured areas, 283 of the 325 paths that look like entity identities resolve; the 42 that
/// do not are spawners, arena props, curse zones and objects/ folders - things that live under
/// Metadata/Monsters/ without being monsters.
///
/// THE @VARIANT SUFFIX IS STRIPPED, from the reference tool rather than from a guess: GameHelper2
/// does the same in MonsterCategories.Get, and its comment says the base equals the MonsterVariety
/// Id. None of the 21 captures carries such a suffix - a preload list holds files, not live
/// entities - so this is the one rule here that the captures could not confirm, and it is kept
/// because the reference is a working tool against the same game.
///
/// WHAT THE QUEST COLUMN IS AND IS NOT. It holds a QuestFlags row number on 68 monsters, and
/// those 68 are exactly the named campaign bosses - Geonor, Jamanra, The Devourer, Ignagduk,
/// Yama The White. So it answers "the thing in front of you is a quest step", not "where is my
/// objective": most objectives are not monsters at all, and the column names no place. The row
/// number is meant for QuestWatch.FlagId, whose table this project already reads from the game's
/// memory - and which it measured at 5717 rows, comfortably above the largest value here (5210).
/// That is a consistency check rather than a proof, and it is the reason the lookup below refuses
/// out-of-range rows instead of showing a flag that belongs to something else.
/// </remarks>
public sealed class MonsterVarieties
{
    /// <summary>Nothing known - what a missing or unreadable table produces.</summary>
    public static MonsterVarieties Empty { get; } = new(
        new Dictionary<string, MonsterVariety>(StringComparer.OrdinalIgnoreCase),
        new Dictionary<int, string>(),
        new Dictionary<int, ModifierMeaning>(),
        new Dictionary<int, string>(),
        new Dictionary<int, MonsterKind>(),
        new Dictionary<int, string>(),
        new Dictionary<int, string>(),
        string.Empty);

    private readonly IReadOnlyDictionary<string, MonsterVariety> _byPath;
    private readonly IReadOnlyDictionary<int, string> _skills;
    private readonly IReadOnlyDictionary<int, ModifierMeaning> _modifiers;
    private readonly IReadOnlyDictionary<int, string> _tags;
    private readonly IReadOnlyDictionary<int, MonsterKind> _types;
    private readonly IReadOnlyDictionary<int, string> _blood;
    private readonly IReadOnlyDictionary<int, string> _resistances;

    private MonsterVarieties(
        IReadOnlyDictionary<string, MonsterVariety> byPath,
        IReadOnlyDictionary<int, string> skills,
        IReadOnlyDictionary<int, ModifierMeaning> modifiers,
        IReadOnlyDictionary<int, string> tags,
        IReadOnlyDictionary<int, MonsterKind> types,
        IReadOnlyDictionary<int, string> blood,
        IReadOnlyDictionary<int, string> resistances,
        string generated)
    {
        _byPath = byPath;
        _skills = skills;
        _modifiers = modifiers;
        _tags = tags;
        _types = types;
        _blood = blood;
        _resistances = resistances;
        Generated = generated;
    }

    /// <summary>When the table was built, as the generator wrote it. Empty when unknown.</summary>
    /// <remarks>
    /// Shown rather than merely stored: a monster table is a snapshot of one patch, and the way
    /// a stale one fails is that new monsters simply have no name - which looks like a lookup bug
    /// rather than an old file.
    /// </remarks>
    public string Generated { get; }

    /// <summary>How many monsters the table knows.</summary>
    public int Count => _byPath.Count;

    /// <summary>Reads the table, or returns <see cref="Empty"/> when it cannot.</summary>
    /// <remarks>Never throws. Without it the entity browser shows paths, as it always did.</remarks>
    public static MonsterVarieties Load(string? path)
    {
        if (path is not { Length: > 0 } || !File.Exists(path))
        {
            return Empty;
        }

        try
        {
            using FileStream stream = File.OpenRead(path);
            MonsterVarietyFile? file =
                JsonSerializer.Deserialize(stream, MonsterJsonContext.Default.MonsterVarietyFile);

            if (file?.Monsters is not { Count: > 0 })
            {
                return Empty;
            }

            var byPath = new Dictionary<string, MonsterVariety>(
                file.Monsters.Count, StringComparer.OrdinalIgnoreCase);

            foreach ((string id, MonsterVariety one) in file.Monsters)
            {
                byPath[Same(id)] = one;
            }

            // THE ROW NUMBER IS THE KEY, kept as the game's own rather than replaced by the
            // name it resolves to. JSON has no integer keys, so the generator writes them as
            // text and they are parsed back here; a row that will not parse is dropped rather
            // than defaulted, because row 0 is a real skill and a silent 0 would name it.
            var skills = new Dictionary<int, string>(file.Skills?.Count ?? 0);
            foreach ((string row, string name) in file.Skills ?? [])
            {
                if (int.TryParse(row, System.Globalization.CultureInfo.InvariantCulture, out int at))
                {
                    skills[at] = name;
                }
            }

            var modifiers = new Dictionary<int, ModifierMeaning>(file.Modifiers?.Count ?? 0);
            foreach ((string row, ModifierMeaning meaning) in file.Modifiers ?? [])
            {
                if (int.TryParse(row, System.Globalization.CultureInfo.InvariantCulture, out int at))
                {
                    modifiers[at] = meaning;
                }
            }

            var tags = new Dictionary<int, string>(file.Tags?.Count ?? 0);
            foreach ((string row, string name) in file.Tags ?? [])
            {
                if (int.TryParse(row, System.Globalization.CultureInfo.InvariantCulture, out int at))
                {
                    tags[at] = name;
                }
            }

            var types = new Dictionary<int, MonsterKind>(file.Types?.Count ?? 0);
            foreach ((string row, MonsterKind kind) in file.Types ?? [])
            {
                if (int.TryParse(row, System.Globalization.CultureInfo.InvariantCulture, out int at))
                {
                    types[at] = kind;
                }
            }

            return new MonsterVarieties(
                byPath,
                skills,
                modifiers,
                tags,
                types,
                Rows(file.Blood),
                Rows(file.Resistances),
                file.Generated ?? string.Empty);
        }
        catch (Exception exception) when (exception is IOException or JsonException or UnauthorizedAccessException)
        {
            return Empty;
        }
    }

    /// <summary>What the table says about the thing at this path, or null.</summary>
    public MonsterVariety? Find(string? entityPath)
    {
        string key = Same(entityPath);
        return key.Length > 0 && _byPath.TryGetValue(key, out MonsterVariety? one) ? one : null;
    }

    /// <summary>How many skills carry a name. Zero when the reference table was not exported.</summary>
    public int NamedSkills => _skills.Count;

    /// <summary>What the game calls this skill row, or empty when it is not known.</summary>
    public string SkillName(int row) => _skills.TryGetValue(row, out string? name) ? name : string.Empty;

    /// <summary>What this modifier row is called and what it sets, or null.</summary>
    public ModifierMeaning? Modifier(int row)
        => _modifiers.TryGetValue(row, out ModifierMeaning? meaning) ? meaning : null;

    /// <summary>
    /// This monster's skills, named where a name is known and numbered where it is not.
    /// </summary>
    /// <remarks>
    /// A ROW WITH NO NAME STAYS VISIBLE, as "#4211" rather than being dropped. A monster with
    /// sixty-seven skills and forty names is a table that needs refreshing; one that quietly
    /// showed forty would look complete and be wrong.
    /// </remarks>
    public IEnumerable<string> Skills(MonsterVariety? one)
    {
        foreach (int row in one?.Effects ?? [])
        {
            string name = SkillName(row);
            yield return name.Length > 0
                ? name
                : "#" + row.ToString(System.Globalization.CultureInfo.InvariantCulture);
        }
    }

    /// <summary>A row-number-keyed map, as the generator wrote it: JSON has no integer keys.</summary>
    /// <remarks>
    /// A row that will not parse is dropped rather than defaulted. Row 0 is a real blood type
    /// (plain Blood, which 1092 monsters carry), so a silent 0 would name it.
    /// </remarks>
    private static Dictionary<int, string> Rows(Dictionary<string, string>? from)
    {
        var got = new Dictionary<int, string>(from?.Count ?? 0);
        foreach ((string row, string name) in from ?? [])
        {
            if (int.TryParse(row, System.Globalization.CultureInfo.InvariantCulture, out int at))
            {
                got[at] = name;
            }
        }

        return got;
    }

    /// <summary>What the game calls this monster's blood, or empty when it is not known.</summary>
    /// <remarks>
    /// Not decoration: NoBlood is why a corpse cannot be raised or exploded, and the 56 rows
    /// separate Blood from BonesNew, GhostBlood, SandBlood and Stone - which is the difference
    /// between a thing that leaves a corpse and one that does not.
    /// </remarks>
    public string BloodName(MonsterVariety? one)
        => one is not null && _blood.TryGetValue(one.Blood, out string? name) ? name : string.Empty;

    /// <summary>
    /// The resistance profiles this monster's TYPE carries, by name.
    /// </summary>
    /// <remarks>
    /// NAMES ONLY, and that is deliberate. Each profile also holds 32 numeric columns -
    /// Fire1..Fire5, Cold1..Cold5 and so on, some of them arrays - and what the numbered tiers
    /// mean is not something the export settles: MinorColdResist reads 30, 30, 30 across the
    /// first three and MajorColdResist reads 75, 60, 50, which fits area tiers and several other
    /// readings equally well. "MajorFireResist" already says what a reader wants; a number here
    /// under a guessed label would be the same mistake as calling AttackSpeed a percentage.
    /// </remarks>
    public IEnumerable<string> ResistancesOf(MonsterVariety? one)
    {
        foreach (int row in Kind(one)?.Resistances ?? [])
        {
            string name = _resistances.TryGetValue(row, out string? got) ? got : string.Empty;
            yield return name.Length > 0
                ? name
                : "#" + row.ToString(System.Globalization.CultureInfo.InvariantCulture);
        }
    }

    /// <summary>How many tags carry a name. Zero when the reference table was not exported.</summary>
    public int NamedTags => _tags.Count;

    /// <summary>
    /// What the game tags this monster as, named where a name is known.
    /// </summary>
    /// <remarks>
    /// This is the column GameHelper2 reads to decide whether a thing is a Beast: humanoid,
    /// human, undead, construct, beast and demon all live here, and a monster may carry several.
    ///
    /// It is NOT a way to tell which league mechanic an area has. The table carries
    /// delve_monster, blight_monster, legion_monster, incursion_monster and breach_monster_fire
    /// among others, and not one monster in the game carries any of them - they are PoE1 rows
    /// left behind, the same trap Data/Balance/PreloadGroups.dat turned out to be. Only sanctum
    /// (77 monsters), expedition (30), azmeri (23), affliction (19), precursor (12) and
    /// sanctified (10) are real.
    /// </remarks>
    public IEnumerable<string> TagsOf(MonsterVariety? one)
    {
        foreach (int row in one?.Tags ?? [])
        {
            string name = Tag(row);
            yield return name.Length > 0
                ? name
                : "#" + row.ToString(System.Globalization.CultureInfo.InvariantCulture);
        }
    }

    /// <summary>What the game calls this tag row, or empty when it is not known.</summary>
    public string Tag(int row) => _tags.TryGetValue(row, out string? name) ? name : string.Empty;

    /// <summary>The defensive block for this monster's type, or null.</summary>
    public MonsterKind? Kind(MonsterVariety? one)
        => one is not null && _types.TryGetValue(one.Type, out MonsterKind? kind) ? kind : null;

    /// <summary>Every modifier this monster carries that resolved to something.</summary>
    /// <remarks>
    /// The three columns are read as one list because nothing here distinguishes them: Mods,
    /// Mods2 and Special_Mods all end up as rows of the same table, and which slot a modifier
    /// arrived in says nothing about what it does.
    /// </remarks>
    public IEnumerable<ModifierMeaning> Modifiers(MonsterVariety? one)
    {
        if (one is null)
        {
            yield break;
        }

        foreach (int row in (one.Mods ?? []).Concat(one.Mods2 ?? []).Concat(one.SpecialMods ?? []))
        {
            ModifierMeaning? meaning = Modifier(row);
            if (meaning is not null)
            {
                yield return meaning;
            }
        }
    }

    /// <summary>
    /// One path in the spelling the table is keyed in.
    /// </summary>
    /// <remarks>
    /// Backslashes become slashes and an @variant suffix is dropped. Both are cheap and both
    /// fail silently when skipped - the symptom is a monster the table plainly contains looking
    /// unknown, which reads as a missing row rather than a mismatched spelling.
    /// </remarks>
    public static string Same(string? path)
    {
        if (string.IsNullOrWhiteSpace(path))
        {
            return string.Empty;
        }

        string cleaned = path.Replace('\\', '/').Trim();
        int variant = cleaned.IndexOf('@', StringComparison.Ordinal);
        return variant >= 0 ? cleaned[..variant] : cleaned;
    }
}

/// <summary>The file's shape.</summary>
internal sealed class MonsterVarietyFile
{
    [JsonPropertyName("generated")]
    public string? Generated { get; init; }

    [JsonPropertyName("monsters")]
    public Dictionary<string, MonsterVariety>? Monsters { get; init; }

    /// <summary>Skill row number to what the game calls it. Row numbers are text - JSON keys.</summary>
    [JsonPropertyName("skills")]
    public Dictionary<string, string>? Skills { get; init; }

    [JsonPropertyName("modifiers")]
    public Dictionary<string, ModifierMeaning>? Modifiers { get; init; }

    [JsonPropertyName("tags")]
    public Dictionary<string, string>? Tags { get; init; }

    [JsonPropertyName("types")]
    public Dictionary<string, MonsterKind>? Types { get; init; }

    [JsonPropertyName("blood")]
    public Dictionary<string, string>? Blood { get; init; }

    [JsonPropertyName("resistances")]
    public Dictionary<string, string>? Resistances { get; init; }
}

/// <summary>Source-generated JSON, so the monster table survives Native AOT.</summary>
[JsonSerializable(typeof(MonsterVarietyFile))]
internal sealed partial class MonsterJsonContext : JsonSerializerContext;

File: scripts/monster-varieties.py
#!/usr/bin/env python3
"""Turn an exported MonsterVarieties table into data/monster-varieties.json.

WHY THIS SCRIPT IS COMMITTED. The table is 124 columns wide and most of them are
dead: Part1_Mods, Part2_Mods, Endgame_Mods and Helmet_ItemVisualIdentity hold
nothing at all across 2734 rows, MonsterArmour holds 16, and several more are
under 1%. Which columns survived is a MEASUREMENT rather than a preference, and
without the script that made the cut the next person has to redo it against a
fresh export to find out whether a missing field was dropped on purpose or lost.

Run it after exporting the tables from the game files:

    python3 scripts/monster-varieties.py MonsterVarieties.csv data/monster-varieties.json \
        --skills GrantedEffects.csv --mods Mods.csv --stats Stats.csv

The three reference tables are OPTIONAL and each one is independent. Without
them the numbers still ship and the browser shows them as numbers, which is what
it did before they were available; with them the same numbers also carry a name.

THE INDICES ARE 0-BASED, and that is measured rather than assumed. An off-by-one
resolves every skill to its neighbour, which is the kind of wrong that reads as
right: a zombie's one skill is "MeleeAtAnimationSpeed" at 1195 and
"MeleeAtAnimationSpeed2" at 1196, and nothing about either looks incorrect. What
settles it is that a named boss's skills carry the boss's own name - Yama gets
GSYamaChaosCloud and YamaSoulrend at 0-based, and DTTPaleFishman at 1-based. The
generator checks that on every run and refuses to write a table that fails it.

WHAT IS STILL NOT RESOLVED. Tags and MonsterType point at tables that are not
here yet, so they stay as numbers. The row number is kept for the resolved
columns too, rather than replaced by the name: it is the game's own identity for
that skill, and a later table keyed on it can still be joined without another
export.
"""

import csv
import json
import sys
from datetime import datetime, timezone

# Columns kept, and the short name each gets in the output. The measurement that
# chose them is in the module docstring; the fill rate of every one of these is
# in the report this script prints, so a future export that guts a column says so
# rather than silently shipping a field of nothing.
NUMBERS = {
    "MonsterType": "type",
    "MovementSpeed": "speed",
    "ObjectSize": "size",
    "ModelSizeMultiplier": "modelSize",
    "MinimumAttackDistance": "minAttack",
    "MaximumAttackDistance": "maxAttack",
    "MinAgroRange": "minAggro",
    "MaxAgroRange": "maxAggro",
    "ExperienceMultiplier": "xp",
    "DamageMultiplier": "damage",
    "LifeMultiplier": "life",
    "AttackSpeed": "attackSpeed",
    "AttackCrit": "crit",
    "BloodType": "blood",
    "Questflag": "quest",
}

# Kept as text. Stance is the one list-shaped column whose values are already
# words ("stance2"), so it needs no second table to mean anything.
TEXT = {
    "Name": "name",
    "Stance": "stance",
}

# Kept as lists of numbers - indices into tables this repository does not have.
INDEX_LISTS = {
    "Tags": "tags",
    "GrantedEffects": "effects",
    "Mods": "mods",
    "Mods2": "mods2",
    "Special_Mods": "specialMods",
}

# Kept as lists of paths.
PATH_LISTS = {
    "InheritsFrom": "inherits",
}

# A float, and the only one.
FLOATS = {
    "PoiseThreshold": "poise",
}


def parts(value):
    """The entries of a bracketed list.

    The export writes lists as ``[a, b]`` with the entries UNQUOTED, so a path
    list reads ``[Metadata/Monsters/X]`` - which is not JSON and not Python. A
    parser that hands those to ast.literal_eval gets an exception, and one that
    swallows the exception reports an empty list for a column that is 26% full.
    That happened while this table was being read, and the wrong answer looked
    exactly like a column nobody fills.
    """
    value = value.strip()
    if not value or value == "[]":
        return []
    if value.startswith("[") and value.endswith("]"):
        value = value[1:-1]
    return [piece.strip() for piece in value.split(",") if piece.strip()]


def number(value):
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def build(rows):
    out = {}
    for row in rows:
        ident = (row.get("Id") or "").strip()

        # "Any" is a sentinel row rather than a monster - it is the only Id in the
        # table that is not a metadata path, and it would never match an entity.
        if not ident.startswith("Metadata/"):
            continue

        one = {}

        for column, short in TEXT.items():
            text = (row.get(column) or "").strip()
            # A name in [brackets] is a placeholder the game shows to nobody
            # ("[ANY MONSTER]"), so it is dropped rather than displayed.
            if text and not text.startswith("["):
                one[short] = text

        for column, short in NUMBERS.items():
            got = number((row.get(column) or "").strip())
            if got is not None:
                one[short] = got

        for column, short in FLOATS.items():
            try:
                one[short] = float((row.get(column) or "").strip())
            except ValueError:
                pass

        for column, short in INDEX_LISTS.items():
            got = [number(p) for p in parts(row.get(column) or "")]
            got = [g for g in got if g is not None]
            if got:
                one[short] = got

        for column, short in PATH_LISTS.items():
            got = parts(row.get(column) or "")
            if got:
                one[short] = got

        # Only when true: 2371 of 2734 are false, and a field that is absent
        # rather than false is 2371 fewer things to write and read.
        if (row.get("BossHealthBar") or "").strip().lower() == "true":
            one["boss"] = True

        base = (row.get("BaseMonsterTypeIndex") or "").strip()
        if base and base != ident:
            one["base"] = base

        out[ident] = one

    return out


def read(path):
    with open(path, encoding="utf-8", errors="replace") as handle:
        return list(csv.DictReader(handle))


# The one mod row that means "this slot is empty". Mods2 is a fixed-width array
# and 1023 of its 3553 references - 29% - point here, so keeping them would put
# five "Nothing" lines under every monster that has one real modifier.
FILLER = "Nothing"


def skill_names(monsters, skills):
    """Row number to skill id, for the rows monsters actually use.

    Shipping all 8347 would be a third again as much file for rows nothing reads.
    """
    wanted = set()
    for one in monsters.values():
        wanted.update(one.get("effects", []))

    named = {}
    for row in sorted(wanted):
        if 0 <= row < len(skills):
            ident = (skills[row].get("Id") or "").strip()
            if ident:
                named[str(row)] = ident
    return named


def prove_alignment(monsters, named):
    """A named boss's skills carry the boss's own name, or the table is misaligned.

    THE CHECK THE WHOLE RESOLUTION RESTS ON. An index shifted by one still
    produces a plausible skill for every monster, so nothing about a wrong table
    looks wrong. What a shift cannot survive is that Ignagduk's skills are all
    called GTIgnagduk-something: the game itself names them after their owner, and
    that is a reference this script cannot talk itself out of.
    """
    checks = [
        ("Metadata/Monsters/YamaBoss/YamaBoss", "Yama"),
        ("Metadata/Monsters/IgnagdukBogWitch/IgnagdukBogWitch", "Ignagduk"),
        ("Metadata/Monsters/MudBurrower/MudBurrowerHeadBoss", "MudBurrower"),
    ]

    for path, word in checks:
        one = monsters.get(path)
        if one is None:
            continue

        got = [named.get(str(row), "") for row in one.get("effects", [])]
        hits = sum(1 for name in got if word.lower() in name.lower())
        if not got:
            continue
        if hits * 2 < len(got):
            raise SystemExit(
                f"skill table looks misaligned: only {hits} of {len(got)} skills on {path} "
                f"mention {word!r}. An off-by-one in the row numbering does exactly this."
            )


def mod_meanings(monsters, mods, stats):
    """Row number to what the modifier is called and which stats it sets."""
    wanted = set()
    for one in monsters.values():
        for key in ("mods", "mods2", "specialMods"):
            wanted.update(one.get(key, []))

    named = {}
    for row in sorted(wanted):
        if not 0 <= row < len(mods):
            continue

        entry = mods[row]
        ident = (entry.get("Id") or "").strip()
        if not ident or ident == FILLER:
            continue

        carried = []
        for slot in range(1, 5):
            at = (entry.get(f"Stat{slot}") or "").strip()
            if not at.isdigit() or int(at) >= len(stats):
                continue

            low, high = 0, 0
            values = parts(entry.get(f"Stat{slot}Value") or "")
            if values:
                low = number(values[0]) or 0
                high = number(values[-1]) if len(values) > 1 else low
            carried.append(
                {"stat": (stats[int(at)].get("Id") or "").strip(), "min": low, "max": high or low}
            )

        named[str(row)] = {"id": ident, "stats": carried}

    return named


def drop_filler(monsters, mods):
    """Take the empty slots out of every monster's modifier lists."""
    if not mods:
        return

    empty = {row for row, entry in enumerate(mods) if (entry.get("Id") or "").strip() == FILLER}
    for one in monsters.values():
        for key in ("mods", "mods2", "specialMods"):
            if key in one:
                kept = [row for row in one[key] if row not in empty]
                if kept:
                    one[key] = kept
                else:
                    del one[key]


def main():
    argv = sys.argv[1:]
    extra = {}
    for flag in ("--skills", "--mods", "--stats"):
        if flag in argv:
            at = argv.index(flag)
            extra[flag[2:]] = argv[at + 1]
            del argv[at : at + 2]

    if len(argv) != 2:
        print(__doc__)
        return 1

    monsters = build(read(argv[0]))

    skills = read(extra["skills"]) if "skills" in extra else []
    mods = read(extra["mods"]) if "mods" in extra else []
    stats = read(extra["stats"]) if "stats" in extra else []

    named_skills = skill_names(monsters, skills) if skills else {}
    if named_skills:
        prove_alignment(monsters, named_skills)

    named_mods = mod_meanings(monsters, mods, stats) if mods and stats else {}
    drop_filler(monsters, mods)

    payload = {
        "generated": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "source": "MonsterVarieties, exported from the game's data tables",
        "skills": named_skills,
        "modifiers": named_mods,
        "monsters": monsters,
    }

    with open(sys.argv[2], "w", encoding="utf-8") as handle:
        json.dump(payload, handle, separators=(",", ":"), sort_keys=True)
        handle.write("\n")

    total = len(monsters)
    print(f"{total} monsters, {len(named_skills)} skills named, {len(named_mods)} modifiers named")
    fields = {}
    for one in monsters.values():
        for key in one:
            fields[key] = fields.get(key, 0) + 1
    for key, count in sorted(fields.items(), key=lambda kv: -kv[1]):
        print(f"  {key:14s} {count:5d}  {100 * count / total:5.1f}%")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
